feat(docs/hover): add support for some html escape sequences

Kasama opened the pull request:
Implements #22757

During handling of textDocument/hover, if the language server returns a payload containing markdown with some html escape sequences, the neovim handler will convert common ones to characters to display in the floating window's buffer.

clason commented:
commit message should be `fix(lsp): convert HTML escape sequences in hover`

(and always better to rebase on master instead of merging master into a PR branch)

Kasama committed:
fix(lsp): convert HTML escape sequences in hover

Implements #22757

During handling of textDocument/hover, if the language server returns a payload
containing markdown with some html escape sequences, the neovim handler will
convert common ones to characters to display in the floating window's buffer.

Signed-off-by: Kasama <robertoaall@gmail.com>

Kasama commented:
> commit message should be `fix(lsp): convert HTML escape sequences in hover`

done

> (and always better to rebase on master instead of merging master into a PR branch -- which includes making PRs from a feature branch, not your fork's master)

I've rebased instead to remove that merge commit. In the future I'll create PRs from a feature branch instead, thanks!

gpanders commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -1378,6 +1378,18 @@ function M.stylize_markdown(bufnr, contents, opts)
 
+  -- Handle some common html escape sequences
+  stripped = vim.tbl_map(function(line)
+    line = string.gsub(line, '&gt;', '>')
+    line = string.gsub(line, '&lt;', '<')
+    line = string.gsub(line, '&quot;', '"')
+    line = string.gsub(line, '&apos;', "'")
+    line = string.gsub(line, '&ensp;', ' ')
+    line = string.gsub(line, '&emsp;', ' ')
+    line = string.gsub(line, '&amp;', '&')
```
This works of course, but this is a good use case for `string.gsub`'s ability to use a table.

```lua
local escapes = {
    ['&gt;'] = '>',
    ['&lt;'] = '<',
    ['&quot;'] = '"',
    ['&apos;'] = "'",
    ['&ensp;'] = ' ',
    ['&emsp;'] = ' ',
    ['&amp;'] = '&',
}

string.gsub(text, '&[^ ;]+;', escapes)
```

Kasama replied:
uuh, nice one. I didn't know gsub could do that. Thanks for the tip. I've updated the PR to use this

mfussenegger reviewed (approved):
Could you also add a test case? Looks like `stylize_markdown` is currently completely untested.
Other than that this looks good to me.

Kasama commented:
> Could you also add a test case? Looks like `stylize_markdown` is currently completely untested. Other than that this looks good to me.

I've added a test file for it, together with a case for my change and some others. Please let me know if it should be somewhere else or behave any differently, I'm not sure where the proper place for it is.

Thanks!

zeertzjq commented on test/functional/plugin/lsp/utils_spec.lua:
```diff
@@ -0,0 +1,74 @@
+
+describe('vim.lsp.util', function()
+  before_each(helpers.clear)
+  after_each(helpers.clear)
```
This is not needed.
```suggestion
```

Kasama replied:
I've removed it, but would you mind clarifying why this is not needed? Is is because other tests already run `helpers.clear` `before_each` test?

zeertzjq replied:
Yes

Kasama replied:
Thanks

justinmk commented on test/functional/plugin/lsp/utils_spec.lua:
```diff
@@ -0,0 +1,75 @@
+      ]], content, opts)
+    end
+
+    it('should stylize markdown code fences', function()
```
"should" is redundant in the test name.

```suggestion
    it('code fences', function()
```

justinmk committed:
Update test/functional/plugin/lsp/utils_spec.lua

justinmk commented on test/functional/plugin/lsp/utils_spec.lua:
```diff
@@ -0,0 +1,75 @@
+      eq(expected, stylize_markdown(lines, opts))
+    end)
+
+    it('should add separator after code block', function()
```
```suggestion
    it('adds separator after code block', function()
```

justinmk committed:
Update test/functional/plugin/lsp/utils_spec.lua

justinmk commented on test/functional/plugin/lsp/utils_spec.lua:
```diff
@@ -0,0 +1,75 @@
+      eq(expected, stylize_markdown(lines, opts))
+    end)
+
+    it('should replace supported HTML entities', function()
```
```suggestion
    it('replaces supported HTML entities', function()
```

justinmk committed:
Update test/functional/plugin/lsp/utils_spec.lua

mjlbach commented:
This should have been a modification to https://github.com/neovim/neovim/blob/master/runtime/syntax/lsp_markdown.vim#L28-L33

I originally implemented the conceal rules so that neovim gui's could handle this with standard markdown rendering libraries.

Now some markup is stripped from the payload, some is concealed, you should decide on a single approach.
